fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathNormalizer.java

```diff
@@ -0,0 +1,200 @@
+        for (ScalarOperator child : scalarOperator.getChildren()) {
+            child.accept(this, context);
+            if (currentPath != null) {
+                allAccessPaths.push(currentPath);
+                currentPath = null;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitVariableReference(ColumnRefOperator variable, Void context) {
+        if (variable.getType().isComplexType()) {
+            currentPath = new AccessPath(variable);
+            allAccessPaths.push(currentPath);
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitSubfield(SubfieldOperator subfieldOperator, Void context) {
+        subfieldOperator.getChild(0).accept(this, context);
+        if (currentPath != null) {
+            subfieldOperator.getFieldNames()
+                    .forEach(p -> currentPath.appendPath(p, TAccessPathType.FIELD));
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitCollectionElement(CollectionElementOperator collectionElementOp, Void context) {
+        if (!collectionElementOp.getChild(1).isConstant()) {
+            collectionElementOp.getChild(1).accept(this, context);
+        }
+
+        collectionElementOp.getChild(0).accept(this, context);
+        if (currentPath == null) {
+            return null;
+        }
+
+        if (!collectionElementOp.getChild(1).isConstant()) {
+            currentPath.appendPath(ColumnAccessPath.PATH_PLACEHOLDER, TAccessPathType.ALL);
+            return null;
+        }
+        currentPath.appendPath(ColumnAccessPath.PATH_PLACEHOLDER, TAccessPathType.INDEX);
+        return null;
+    }
+
+    @Override
+    public Void visitCall(CallOperator call, Void context) {
+        if (!PruneSubfieldRule.SUPPORT_FUNCTIONS.contains(call.getFnName())) {
+            return visit(call, context);
+        }
+
+        if (call.getFnName().equals(FunctionSet.MAP_KEYS)) {
+            call.getChild(0).accept(this, context);
+            if (currentPath != null) {
+                currentPath.appendPath(ColumnAccessPath.PATH_PLACEHOLDER, TAccessPathType.KEY);
+            }
+            return null;
+        } else if (FunctionSet.MAP_SIZE.equals(call.getFnName())
+                || FunctionSet.CARDINALITY.equals(call.getFnName())
+                || FunctionSet.ARRAY_LENGTH.equals(call.getFnName())) {
+            call.getChild(0).accept(this, context);
+            if (currentPath != null) {
+                currentPath.appendPath(ColumnAccessPath.PATH_PLACEHOLDER, TAccessPathType.OFFSET);
+            }
+        }
+
+        return null;
+    }
+
+    @Override
+    public Void visitIsNullPredicate(IsNullPredicateOperator predicate, Void context) {
+        predicate.getChild(0).accept(this, context);
+        if (currentPath != null) {
+            if (predicate.getChild(0).getType().isMapType()) {
+                currentPath.appendPath(ColumnAccessPath.PATH_PLACEHOLDER, TAccessPathType.OFFSET);
```

Comment:
judge is null according to offset? offset difference is 0 means is null? but that is not true

Reply:
judge null by null-column, not offset, set offset just reuse flag, don't need a new flag to mark only need null-column